### ISSUE TITLE
Spoofing of organisations possible through any CSP node

### ISSUE THREAD

**Issue by iglocska** [melicertes/csp#27]
Steps to reproduce:

(- Get any sharing group via the issue described in the previous MISP sync issue)  => possibly not needed
- replace your organisation's UUID with that of the organisation you wish to spoof
- MISP's protection against this sort of spoofing is circumvented by CSP
- CSP will check the current state of the organisations to decide whether data can be propagated to the node

**Comment by iglocska** [comment]
Confirmed: pre-condition is not needed.

**Comment by intrasoft-rmalik** [comment]
@iglocska thanks for this. We will investigate this further.

**Comment by thanosa75** [comment]
Further info needed; @iglocska how are you going to change UUIDs? are you somehow modifying the payload of the sync operation? can you provide a test case we can reproduce the issue ?

**Comment by iglocska** [comment]
Administration - list organisations - edit in MISP.

**Comment by thanosa75** [comment]
Hi,
some analysis below on this:
an exploit would roughly entail:
- A malevolent MISP administrator to change the sharing group data ON PURPOSE
- A malevolent user (could be the malevolent administrator) sharing 'bad data' (not sure what that would look like though)
- A receiver side where all input validations (also those in MISP!) would NOT handle the 'bad data'
- A receiver side where the 'bad data' would compromise anything OUTSIDE the receiving MISP, e.g. by re-sharing to a non-MeliCERTes MISP ON PURPOSE (because a CSIRT would be foolish to not have some triage between MeliCERTes and non-MeliCERTes-world systems, I think?).
Please also see below the FIRST.org calculation for the vulnerability assesment.
![GitHub 27 CVSS-SIG FIRST-ORG](https://user-images.githubusercontent.com/1307607/66394064-b9c0ad80-e9dc-11e9-83c0-7d52e9096fce.PNG)


**note:  priority in accordance with a CVSS-SIG score of 5.8, MEDIUM **

**Comment by bartvanriel** [comment]
After some exploration , we have come to the following scenario which could constitute a business risk:
+++
Lets's define 3 CSPs: 
•	V -> the victim's CSP
•	A -> the attacker's CSP
•	S -> the CSP which the attacker claims to be (the name used for spoofing).

The business impact of Option 2 would be that V will receive a message from A, thinking it originates from S.

So if V trusts S it will take that message contents as trustworthy and act upon it. In case of the message contents being an (faked) IOC relevant to the Victim's premises, that might lead to the following: 
•	the alteration or shutdown of services, 
•	possible information disclosure by sharing IOC-related data with  others as thought needed to further mitigate the risk

The real-world business risk of this scenario depends on the local setup in which Melicertes is used. For a CSP which facilitates the core services of a national CERT the impact may be significant.
+++
But the above scenario only works if changing the ‘Sharing Group’ CSP-ID-‘derivative’ in the ‘organizations’ settings of MISP (the part that requires Admin access) actually means that the ‘fake’ CSP-ID-‘derivative’ actually can be received as the ‘originating CSP ID’ by receiving CSPs. 
And that can’t happen. Because at the incoming CSP side, the CSP-ID in the envelope is checked against the CSP-ID of the sending CSP, based on SSL Certificate Headers (CSP-to-CSP authentication is certificate based). The certificates are not changeable since an individual CSIRT does not control the PKI of the root CA. So ‘spoofed’ data cannot be effectively shared with other CSPs, which mitigates the spoofing.

**Comment by bartvanriel** [comment]
Recalculation of the CVSS score from the first.org calculator. 
Explanation of the parameters chosen:
AV - Local since changing sharing group parameters requires MeliCERTes MISP access, which is not public, but only within the local CSIRTs environment.
AC - High since it requires quite a few steps
PR - High since elevated privileges are required to change sharing group parameters
UI - Required since you need access to the actual MISP application
S - UNCHANGED since exploit cannot affect other entities (other MeliCERTes instances will not accept the shared data because of the mismatch between spoofed CSPID and CSPID from the Certificate Authentication between nodes)
C - Low because this only changes a sharing group setting which does not propagate in any way beyond the local MISP and therefore cannot change the availability of Confidential data, not even in the local MISP (does not alter access rights for example)
I - Low since the integrity of the exploited MISP instance is not severely damaged in the sense that the whole MISP is corrupted. Some data records are affected locally but not the whole MISP. 
A - None since changing sharing group parameters does not render the MISP unavailable at al. Actually, MISP continues to work as normal.

The score now comes to 1.8 LOW. Most significant change from the previous calculation is the fact that Confidentiality is 'Low' since no MISP information can be falsely propagated to other CSIRTs via MeliCERTes, because of consistency validations at receiving MeliCERTes nodes.
![GitHub 27 updated score](https://user-images.githubusercontent.com/5075734/74133402-c81c1180-4be8-11ea-88a3-664954d9a8fb.PNG)

**Comment by iglocska** [comment]
We have demonstrated it during the training, the impact is obviously much higher. Via a timing attack we could abuse this to receive information meant for trust circle members that the malicious user was not involved in. 

Confidentiality should be high - you are receiving data not meant for you and are breaking the trust placed in trust circles only sending data to the addressed recipient. 

Attack complexity is low, all it required was for us to delete the target organisation's object and modify our own organisation's UUID before the organisation could get repopulated.

Attack vector is not local but network, it's a different Node that we're getting the data from.

Infrastructure used was the consortium's training infrastructure with 3 player nodes and was shown live to 2 members of the consortium. If you have fixed it since, I'd love to see the commit that resolves this.

**Comment by thanosa75** [comment]
Hello @iglocska @bartvanriel 

We have narrowed down the root cause of this issue to a bug discovered summer 2019 and resolved a bit later than that. I will point out the relevant git commits in the bottom of this update, but the 'gist' of this is: Internally the routing component (Shared Services - IL) is using AMQ Queues, and events incoming or outgoing are dispatched via the relevant queue for the external CSPs. We were made aware of a situation that a specially crafted (which in reality turned out not really 'specially' crafted after all) CSP ID (e.g name of the CSP) in the TC application could end up creating multiple queues (part of the name could be interpreted as a wildcard by AMQ) as it was not sanitised either at the time of entry or at the time of use (in AMQ). This has been fixed in the relevant commits:

TC: https://github.com/melicertes/csp/commit/7fada164b931ee0cc7555da5bd87397cb4e885d6#diff-0beb5cf155269635f519b9f3c12b8bc4R156 (js function short_name_is_valid)
IL: https://github.com/melicertes/csp/commit/dd15d0c9078553044f37536baf60b580a655664e#diff-cb8f295b42a57b02f4c3c808c17410fcR51 (safeQueueName)

In our tests we cannot replicate the situation anymore and these fixes are part of the latest release delivered in early December to the CENTRAL by the Consortium.

Best regards,
Thanos
